backend/src/main/scala/com/softwaremill/adopttapir/starter/api/StarterRequestValidator.scala

```diff
@@ -0,0 +1,133 @@
+  }
+
+  case class GroupIdShouldFollowJavaPackageConvention(input: String) extends RequestValidation {
+    override val errMessage: String = s"GroupId: `$input` should follow Java package convention"
+  }
+
+  abstract class EffectValidation {
+    val effect: EffectRequest
+    val implementation: ServerImplementationRequest
+    protected val prefixMessage = s"Picked $effect with $implementation -"
+  }
+
+  case class FutureEffectWillWorkOnlyWithAkkaAndNetty(effect: EffectRequest, implementation: ServerImplementationRequest)
+      extends EffectValidation
+      with RequestValidation {
+    override val errMessage: String = s"$prefixMessage Future effect will work only with Akka and Netty"
+  }
+
+  case class IOEffectWillWorkOnlyWithHttp4sAndNetty(effect: EffectRequest, implementation: ServerImplementationRequest)
+      extends EffectValidation
+      with RequestValidation {
+    override val errMessage: String = s"$prefixMessage IO effect will work only with Http4 and Netty"
+  }
+
+  case class ZIOEffectWillWorkOnlyWithHttp4sAndZIOHttp(effect: EffectRequest, implementation: ServerImplementationRequest)
+      extends EffectValidation
+      with RequestValidation {
+    override val errMessage: String = s"$prefixMessage ZIO effect will work only with Http4s and ZIOHttp"
+  }
+}
+
+sealed trait FormValidator {
+  type ValidationResult[A] = ValidatedNec[RequestValidation, A]
+
+  def validate(tapirVersion: String, r: StarterRequest): Either[IncorrectInput, StarterDetails] =
+    (
+      validateSemanticVersioning(tapirVersion),
+      validateProjectName(r.projectName),
+      validateGroupId(r.groupId),
+      validateEffectWithImplementation(r.effect, r.implementation)
+    ).mapN { case (tapirVersion, projectName, groupId, (effect, serverImplementation)) =>
+      effect match {
+        case EffectRequest.IOEffect     => IOStarterDetails(projectName, groupId, serverImplementation.toModel(), tapirVersion)
+        case EffectRequest.FutureEffect => FutureStarterDetails(projectName, groupId, serverImplementation.toModel(), tapirVersion)
+        case EffectRequest.ZIOEffect    => ZIOStarterDetails(projectName, groupId, serverImplementation.toModel(), tapirVersion)
+      }
+    }.leftMap(errors => IncorrectInput(errors.toNonEmptyList.map(_.errMessage).toList.mkString(System.lineSeparator())))
+      .toEither
+
+  private def validateSemanticVersioning(version: String): ValidationResult[String] = {
+    // regex from: https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
+    val semverRgx =
+      "^(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$"
+    if (version.matches(semverRgx))
+      version.valid
+    else NotInSemverNotation(version).invalidNec
+  }
+
+  private def validateProjectName(projectName: String): ValidationResult[String] = {
```

Comment:
I think it should also not contain any "weird" chars such as ą, ę etc. So validating `[a-z0-9]+` should do the trick :)